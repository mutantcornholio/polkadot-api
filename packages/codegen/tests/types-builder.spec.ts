import { expect, describe, it, beforeAll } from "vitest"
import fs from "fs/promises"
import assert from "assert"

import { defaultDeclarations, getTypesBuilder } from "@/types-builder"
import { MetadataLookup } from "@polkadot-api/metadata-builders"
import { V14, V15, metadata } from "@polkadot-api/substrate-bindings"
import { getLookupFn } from "@polkadot-api/metadata-builders"
import knownTypes from "@/known-types"
import { getChecksumBuilder } from "@polkadot-api/metadata-builders"
import { getDispatchErrorId } from "@/generate-descriptors"

describe("types-builder", () => {
  let metadataLookup: MetadataLookup
  let metadataDecoded: V14 | V15
  let checksumBuilder: ReturnType<typeof getChecksumBuilder>

  const getBuilder = () =>
    getTypesBuilder(
      defaultDeclarations(),
      metadataLookup,
      knownTypes,
      checksumBuilder,
    )

  beforeAll(async () => {
    const metadataRaw = await fs.readFile("./tests/ksm.bin")
    metadataDecoded = metadata.dec(metadataRaw).metadata.value as V14 | V15
    metadataLookup = getLookupFn(metadataDecoded)
    checksumBuilder = getChecksumBuilder(metadataLookup)
  })

  describe("buildTypeDefinition", () => {
    it("should generate correct dispatchErrorType", () => {
      const typesBuilder = getBuilder()

      const dispatchErrorId = getDispatchErrorId(metadataLookup)!
      expect(typesBuilder.buildTypeDefinition(dispatchErrorId)).toEqual(
        "Anonymize<I69ftqh8hqffme>",
      )
    })
  })

  describe("buildStorage", () => {
    for (const [pallet, entry, expectedKey, expectedVal] of [
      ["System", "Account", "[Key: SS58String]", "Anonymize<I5sesotjlssv2d>"],
      ["System", "Events", "[]", "Anonymize<Iarv2l5iokakqs>"],
      ["XcmPallet", "VersionNotifiers", "Anonymize<I5dnj7v75alj6b>", "bigint"],
      ["XcmPallet", "VersionDiscoveryQueue", "[]", "Anonymize<I61487j17uoqo8>"],
    ]) {
      it(`generates ${expectedKey}: ${expectedVal} for ${pallet}::${entry}`, () => {
        const typesBuilder = getBuilder()

        const result = typesBuilder.buildStorage(pallet, entry)
        expect(result.key).toEqual(expectedKey)
        expect(result.val).toEqual(expectedVal)
      })
    }
  })

  describe("buildEvent", () => {
    for (const [pallet, entry, expected] of [
      ["System", "ExtrinsicFailed", "Anonymize<Iau4actk6tfs21>"],
      ["ConvictionVoting", "Undelegated", "Anonymize<SS58String>"],
    ]) {
      it(`generates ${expected} for ${pallet}::${entry}`, () => {
        const typesBuilder = getBuilder()

        const result = typesBuilder.buildEvent(pallet, entry)
        expect(result).toEqual(expected)
      })
    }
  })

  describe("buildError", () => {
    for (const [pallet, entry, expected] of [
      ["XcmPallet", "InvalidAssetNotConcrete", "undefined"],
    ]) {
      it(`generates ${expected} for ${pallet}::${entry}`, () => {
        const typesBuilder = getBuilder()

        const result = typesBuilder.buildError(pallet, entry)
        expect(result).toEqual(expected)
      })
    }
  })

  describe("buildCall", () => {
    for (const [pallet, entry, expected] of [
      ["System", "authorize_upgrade", "Anonymize<Ib51vk42m1po4n>"],
      ["Society", "resign_candidacy", "undefined"],
    ]) {
      it(`generates ${expected} for ${pallet}::${entry}`, () => {
        const typesBuilder = getBuilder()

        const result = typesBuilder.buildCall(pallet, entry)
        expect(result).toEqual(expected)
      })
    }
  })

  describe("buildRuntimeCall", () => {
    for (const [pallet, entry, expectedArgs, expectedValue] of [
      ["Core", "version", "[]", "Anonymize<Ic6nglu2db2c36>"],
      [
        "Core",
        "execute_block",
        "[block: Anonymize<Iaqet9jc3ihboe>]",
        "undefined",
      ],
    ]) {
      it(`generates ${expectedArgs} -> ${expectedValue} for ${pallet}::${entry}`, () => {
        const typesBuilder = getBuilder()

        const result = typesBuilder.buildRuntimeCall(pallet, entry)
        expect(result.args).toEqual(expectedArgs)
        expect(result.value).toEqual(expectedValue)
      })
    }
  })

  describe("buildConstant", () => {
    for (const [pallet, entry, expected] of [
      ["System", "BlockHashCount", "number"],
      ["System", "Version", "Anonymize<Ic6nglu2db2c36>"],
    ]) {
      it(`generates ${expected} for ${pallet}::${entry}`, () => {
        const typesBuilder = getBuilder()

        const result = typesBuilder.buildConstant(pallet, entry)
        expect(result).toEqual(expected)
      })
    }
  })

  describe("imports", () => {
    it("collects imports from both common types and papi itself", () => {
      const typesBuilder = getBuilder()

      metadataDecoded.pallets.forEach((pallet) => {
        pallet.storage?.items.forEach(({ name }) =>
          typesBuilder.buildStorage(pallet.name, name),
        )
        pallet.constants?.forEach(({ name }) =>
          typesBuilder.buildConstant(pallet.name, name),
        )

        const forEachVariant = (
          type: "calls" | "errors" | "events",
          cb: (name: string) => void,
        ) => {
          if (!pallet[type]) return
          const lookup = metadataDecoded.lookup[pallet[type]!]
          assert(lookup.def.tag === "variant")
          lookup.def.value.forEach(({ name }) => {
            cb(name)
          })
        }
        forEachVariant("calls", (name) =>
          typesBuilder.buildCall(pallet.name, name),
        )
        forEachVariant("errors", (name) =>
          typesBuilder.buildError(pallet.name, name),
        )
        forEachVariant("events", (name) =>
          typesBuilder.buildEvent(pallet.name, name),
        )
      })

      expect(typesBuilder.getTypeFileImports()).toMatchSnapshot()
      expect(typesBuilder.getClientFileImports()).toMatchSnapshot()
    })
  })
})
