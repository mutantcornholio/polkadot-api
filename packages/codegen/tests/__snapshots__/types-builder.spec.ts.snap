// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`types-builder > imports > collects imports from both common types and papi itself 1`] = `
[
  "I5sesotjlssv2d",
  "Iffmde3ekjedi9",
  "I2h3virg0mao09",
  "I4mddgoa69c0a2",
  "Ia5dve0uamet4h",
  "I95g6i7ilua7lq",
  "Ieniouoqkq4icf",
  "Phase",
  "Ibgl04rn6nbfm6",
  "In7a38730s6qs",
  "If15el53dd76v9",
  "I9s0ave7t0vnrk",
  "Ic6nglu2db2c36",
  "I8ofcg5rbj0g2c",
  "I4adgbll7gku4i",
  "I6pjjpfvhvcfru",
  "I9pj91mj79qekl",
  "I39uah9nss64h9",
  "Ik64dknsq7k08",
  "Ib51vk42m1po4n",
  "Ia82mnkmeo2rhc",
  "Iau4actk6tfs21",
  "Icbccs0ug47ilf",
  "I855j4i3kr8ko1",
  "I3geksg000c171",
  "BabeDigestsNextConfigDescriptor",
  "Ic5m5lp1oioo8r",
  "Idq7or56ds2f13",
  "I4s6vifaf8k998",
  "I9jd27rnpm8ttv",
  "I8jnd4d8ip6djo",
  "Ifip05kcrl65am",
  "I50ppnqasq4tjq",
  "I9fin09kkg0jaj",
  "Idcr6u6361oad9",
  "Iff9heri56m1mb",
  "I666bl2fqjkejo",
  "Ifrr0vsnqf33qj",
  "I6v36c5qqol9lm",
  "Ia1u3jll6a06ae",
  "I1q8tnt1cluu5j",
  "I8ds64oj6581v0",
  "Ia7pdug7cdsg8g",
  "Ifosl32ihgoeq6",
  "I2l1ctuihi2mfd",
  "I43m56mhq8l63",
  "Ibag9pvfa92nol",
  "Ifs3v4d6ugmts1",
  "I91k5co59ruigu",
  "Ibmr18suc9ikh9",
  "I7dnev3etp2g3m",
  "I5u8olqbbvfnvf",
  "Icv68aq8841478",
  "Ic262ibdoec56a",
  "Iflcfm9b6nlmdd",
  "Ijrsf4mnp3eka",
  "Id5fm4p8lj5qgi",
  "I8tjvj9uq4b7hi",
  "I4cbvqmqadhrea",
  "I3qt1hgg4djhgb",
  "I4fooe9dun9o0t",
  "TransactionPaymentReleases",
  "Ier2cke86dqbr2",
  "Ia2lhg7l2hilo3",
  "Ic12aht5vh2sen",
  "StakingRewardDestination",
  "I9o7ssi9vmhmgr",
  "Ic3m9d6tdl6gi2",
  "Ib3j7gb0jgs38u",
  "Ifekshcrgkl12g",
  "I7svnfko10tq2e",
  "I6flrronqs3l6n",
  "I97fulj5h3ik95",
  "Ia8896dq44k9m4",
  "Icgljjb6j82uhn",
  "Iff9p3c7k6pfoi",
  "StakingForcing",
  "Iafq6t4rgheait",
  "I4ojmnsk1dchql",
  "Iinkhfdlka9ch",
  "I2kj4j6mp68hf8",
  "I6ouflveob4eli",
  "Ia9ek8s49lerat",
  "I2eip8tc75dpje",
  "I564va64vtidbq",
  "Ie5v6njpckr05b",
  "I328av3j0bgmjb",
  "I4tuqm9ato907i",
  "Iba37vq6vinim6",
  "I9dgmcnuamt5p8",
  "I3vh014cqgmrfd",
  "Ifhs60omlhvt3",
  "If34udpd5e57vi",
  "I39t01nnod9109",
  "Ie5vbnd9198quk",
  "I3h6murn8bd4v5",
  "I6k6jf8ncesuu3",
  "I57m3u63hnlsnv",
  "I46cnl4ibg49mg",
  "Idl3umm12u5pa",
  "I5ont0141q9ss5",
  "Ie6j49utvii126",
  "I3v6ks33uluhnj",
  "I3kiiim1cds68i",
  "I4k60mkh2r6jjg",
  "I1au3fq4n84nv3",
  "Iejaj7m7qka9tr",
  "Idnak900lt5lm8",
  "I27n7lbd66730p",
  "I2hq50pu2kdjpo",
  "Ifk8eme5o7mukf",
  "Iau4cgm6ih61cf",
  "I6ir616rur362k",
  "Ic19as7nbst738",
  "I54umskavgc9du",
  "I2ip7o9e2tc5sf",
  "I5egvk6hadac5h",
  "I2bqvqrg0sbrdj",
  "I23nq3fsgtejt",
  "Iempvdlhc5ih6g",
  "I4pact7n2e9a0i",
  "Idt624nf41g34e",
  "I9kr8cseidc66h",
  "I82jm9g7pufuel",
  "I5oi8saufice6j",
  "GrandpaStoredState",
  "I7pe2me3i3vtn9",
  "I7ne83r38c2sqq",
  "I2hviml3snvhhn",
  "I5768ac424h061",
  "Iegmj7n48sc3am",
  "I5v9a2mdqq4t8u",
  "I35p85j063s0il",
  "Iaa58nmghldi9a",
  "I5uob9vhdv4sl4",
  "Icm9m0qeemu66d",
  "Iapv3kiqmahooe",
  "Ibv74ul0c0qf0f",
  "I44hc4lgsn4o1j",
  "I8iksqi3eani0a",
  "I16enopmju1p0q",
  "Ifgqhle2413de7",
  "I43kq8qudg7pq9",
  "I76riseemre533",
  "I38bmcrmh852rk",
  "I4hcillge8de5f",
  "If52u7u3r3ve0n",
  "Iek7v4hrgnq6iv",
  "ConvictionVotingVoteVoting",
  "If9jidduiuq7vv",
  "Idnsr2pndm36h0",
  "Icmqhu5ghlej7a",
  "I8steo882k7qns",
  "I1jitumdc075ot",
  "I5f178ab6b89t3",
  "I2is3ho2fhk7fm",
  "I2na29tt2afp0j",
  "I85t6l9pssiov4",
  "Ibafpkl9hhno69",
  "I4dvjo4haia091",
  "Icbio0e1f0034b",
  "I8c0vkqjjipnuj",
  "I229ijht536qdu",
  "I62nte77gksm0f",
  "I9cg2delv92pvq",
  "Ilhp45uime5tp",
  "I4f1hv034jf1dt",
  "I5g2vv0ckl2m8b",
  "I3gg47bgkgq9tr",
  "I2jtc3igaca648",
  "I9fm9gk5nkilj2",
  "I8bvk21lpmah75",
  "I449n3riv6jbum",
  "I36c8spt0udrvh",
  "Im1pm2vf6llcn",
  "I21jsoeb0o6476",
  "Ier6ck0tpfo7",
  "Icousgkbefm7ke",
  "Ic6ecdcp9ut7jd",
  "I27notaksll8qt",
  "I1adbcfi5uc62r",
  "Ibf6ucefn8fh49",
  "I3i3acbv6b2h9j",
  "Ibbj9ruloghdfg",
  "I44q43ie1r39o2",
  "I2phecamkn3pej",
  "ClaimsStatementKind",
  "I6uag8j5aql8q",
  "Isq3k9rh2c0l4",
  "I1dqiovk0tpoah",
  "I1ntko0oih7v1a",
  "I2tf5qmg09624f",
  "Ie3hcrrq6r18fs",
  "Ira2igbhpnuo7",
  "Ia97lhcik62qhr",
  "I3au22oqgrc8fr",
  "Ied2hngepclu99",
  "I89kp1gbpgfia2",
  "Idvnn1v30725f8",
  "Iarmkett7qcmm5",
  "I9hkvi4k3kh916",
  "Iam8qhv8b3jn4n",
  "I1659sf24o6ga8",
  "Iehiasr6uoqbq",
  "I8adv0j4q3uu5d",
  "I7d29ossbog0d3",
  "I7mgf69gvksvu2",
  "I3sfg6klg56vmk",
  "If1csk2jn0j9qv",
  "I3fe77rqmi4k9q",
  "I26anifrfnljoi",
  "I5118nq7gnqhm8",
  "I7lt48co2f6pcd",
  "I4b66js88p45m8",
  "Ieh40op1lafir9",
  "Icjp7f9u9fno72",
  "I8ircspqpe79k7",
  "I1vlt8l6nvnen6",
  "I72vf0j9juc1df",
  "Ie0phfcvllr4od",
  "I476ep66p6ddg1",
  "Ie3gphha4ejh40",
  "If0btdsid69rrm",
  "Ictusacdcqj61u",
  "Ibsolu3c3q1lb4",
  "Ib116djgo9u4uo",
  "Ibprd8oi8phm62",
  "Idlqqo993i780l",
  "I7cn5k1k5qthcf",
  "Ievmma79msp872",
  "I6s6ihmfj6j5qq",
  "I117g5egc6i2d5",
  "Id6s15tsehsnp8",
  "I9vkkue6cq74et",
  "Ibu56t5h1q49i4",
  "I1dmtl5t34b9g",
  "Ifble4juuml5ig",
  "Version",
  "Ibfjtjhevlfun",
  "Ih2guv8q34bc1",
  "I7k4ereorluoah",
  "Ict9ivhr2c5hv0",
  "I4cl5oaa43djc7",
  "Ievr89968437gm",
  "Ic836nv7r3fq4i",
  "I4q39t5hn830vp",
  "I6mr1qs0ff7sss",
  "I5n4sebgkfr760",
  "Ibt2srb3203f39",
  "Ifs1i5fk9cqvr6",
  "Ibobo8b59ui7or",
  "Ib8vp5vpksivuf",
  "I5duf3b0i0i3du",
  "Ienusoeb625ftq",
  "I64klrfql92glr",
  "I9p9lq3rej5bhc",
  "I3enkdnimcn74o",
  "I1ogpf1ej1kelg",
  "I5ct68gln4v223",
  "I8njjqaantc0h7",
  "I5d7pa7lfklkud",
  "I1ai4pouuusmgg",
  "Idfnchp2o9tp75",
  "I9sm5v527tn4c6",
  "I2ur0oeqg495j8",
  "I38c7pd0e7js3p",
  "Iag146hmjgqfgj",
  "I8uo3fpd3bcc6f",
  "Ifku16nqqp96nr",
  "Idf4chfb9beoqk",
  "Ideaemvoneh309",
  "I3d9o9d7epp66v",
  "Iep27ialq4a7o7",
  "Iasu5jvoqr43mv",
  "I28r95gak1ju7a",
  "I5qolde99acmd1",
  "PreimageOldRequestStatus",
  "PreimageRequestStatus",
  "I82nfqfkd48n10",
  "I1jm8m1rh9e20v",
  "I3o5j3bli1pd8e",
  "I8phqps8r3of7e",
  "I2a839vbf5817q",
  "Ia9p5bg6p18r0i",
  "Ide9116k1d76l7",
  "Iel2dba701u54t",
  "I90n6nnkpdahrh",
  "Id9idaj83175f9",
  "Ie1semicfuv5uu",
  "If25fjs9o37co1",
  "I70sc1pdo8vtos",
  "Ibofbvvaehln4e",
  "I8mk5kjgn02hi8",
  "I77qn9tu99mv6h",
  "I2gr10p66od9ch",
  "I6tv2j51i0cd22",
  "I60p8l86a8cm59",
  "I3m3sk2lgcabvp",
  "I5pf572duh4oeg",
  "ElectionProviderMultiPhasePhase",
  "Ictkaqdbfabuek",
  "Ia7o65280hur3p",
  "Iasd2iat48n080",
  "Ic8d01sg6acf60",
  "I84stt310eb6c3",
  "I8s6n43okuj2b1",
  "Ic50i89t5eabvg",
  "I80q14um2s2ckg",
  "I5qs1t1erfi7u8",
  "I2ou0mtsgm3l71",
  "Ifsme8miqq9006",
  "I4mol6k10mv0io",
  "Iec90vukseit9e",
  "I7j4m7a3pkvsf4",
  "Ic2n50kpnu5mae",
  "I2cqtnj49ogo81",
  "I7btpfdeli5ea8",
  "Ic79q40j9qptgh",
  "I6cs1itejju2vv",
  "Icckjru8f8flpj",
  "Iba38lgbjrs7je",
  "Ie2vmgag6qiue4",
  "I1kpuet4lkcdpf",
  "I5tr1gp2t9i6oh",
  "I7ls1r4u61csif",
  "I3gjgtso69eopr",
  "I9bin2jc70qt6q",
  "Ic5t26f9cp3tvk",
  "I39k39h6vu4hbq",
  "Iafqnechp3omqg",
  "Ibrlrqnm294npo",
  "Id0haetkq9cdrn",
  "I1p98btdd62l70",
  "BagsListListListError",
  "I37454vatvmm1l",
  "Iblau1qa7u7fet",
  "Idphjddn2h69vc",
  "Idhh9vuu2bderg",
  "If6qa32dj75gu1",
  "I7oo2mprv1qd1s",
  "NominationPoolsClaimPermission",
  "Ieg1oc56mamrl5",
  "I2vu5vj7173ik9",
  "Iekrjqn2dsea54",
  "I36uoc8t9liv80",
  "I4f7s38fnc5uj8",
  "If3rcp19d58ctc",
  "I4r3gtp044pif4",
  "I47a2tsd2o2b1c",
  "Ifc9k1s0e9nv8e",
  "I4ihj26hl75e5p",
  "I2dl8ekhm2t22h",
  "I13us5e5h5645o",
  "I931cottvong90",
  "Indc2esbhpfa0",
  "I1ors0vru14it3",
  "I40s11r8nagn2g",
  "I6bjj87fr5g9nl",
  "I8cbluptqo8kbp",
  "I81cc4plffa1dm",
  "I3ihan8icf0c5k",
  "NominationPoolsPalletDefensiveError",
  "I1ti389kf8t6oi",
  "If4nnre373amul",
  "I55kbor0ocqk6h",
  "Idsj9cg7j96kpc",
  "Ido4u9drncfaml",
  "Ie8c7ctks8ur2p",
  "I7vqogd77mmdlm",
  "I6mik29s5073td",
  "I2m0sqmb75cnpb",
  "I49agc5b62mehu",
  "Iatq9jda4hq6pg",
  "I2g87evcjlgmqi",
  "I2eh80qovrl7h2",
  "I9j0ul7nh7b8jv",
  "I8vevhfg3sioic",
  "Ic0he9tlf9ll0u",
  "I49is1a3dbqc3s",
  "I3uqicpe83t965",
  "Id581arok0b1nj",
  "I9jsikd1ghmc7l",
  "I2f6mha3v4ooda",
  "Iasqjdhasi408s",
  "I6krn2lsleo87n",
  "Iaid4btmkr5thp",
  "I27il479s8gsv0",
  "I78e99376i3249",
  "I2jernj9ccmd7t",
  "Ic1d4u2opv3fst",
  "Icdu1nb48l753j",
  "I618m2uag0aopg",
  "Ic8i89mfkmn3n7",
  "Ilg2fqs6cjgs3",
  "Ibbf0j7ol2abrv",
  "Iarlj3qd8u1v13",
  "I2ng2krd94ceva",
  "Ie18mk3nmrn3nr",
  "Iap8hdu530i2rj",
  "ParachainsParasParaLifecycle",
  "I79cs1p3m59mo7",
  "UpgradeGoAhead",
  "UpgradeRestriction",
  "I2duhnt686rv0q",
  "I1k3urvkqqshbc",
  "I2ff0ffsh15vej",
  "I1orfg86bkg123",
  "Iaus4cb3drhu9q",
  "Ivnsat10lv9d6",
  "Ibncli8qttt2c2",
  "I33rft6ag34efs",
  "I9tmok5kceg2bg",
  "I7ulu3h1ibu60i",
  "I85icj2qbjeqbe",
  "I6ljjd4b5fa4ov",
  "Ibhmrlkcu01imb",
  "I50mrcbubp554e",
  "Id43g4eveajpkl",
  "I7iua3ehrgl4va",
  "Iev3u09i2vqn93",
  "I9olhgo2o08h7b",
  "Ibuhbp68e6tkct",
  "Idrevppfiubhve",
  "I9s2h36kr71vk9",
  "I4lkbiubo9ogq9",
  "Id1baei7m8gkhk",
  "I96ftepqm4vs7m",
  "Ic3430470j4mbv",
  "Ic3n7nqb6fffo0",
  "Id2bej717ckub0",
  "I545vo2e86o5i4",
  "I9m4rd2a7lc9md",
  "I80rnntpog8qp6",
  "I87u7jalc0lhah",
  "I4p5t2krb1gmvp",
  "I4arjljr6dpflb",
  "I3i09nus3ku37s",
  "I2e447aa6a0imh",
  "I5kqchhvguhfvt",
  "I437u7rqtshfms",
  "Idgino88phkfa9",
  "Iaa7g3f5tlv3gf",
  "I89dbspj67f63r",
  "Ia7efn5eqruldn",
  "I9dasmua8326io",
  "I3g90iebhds6kb",
  "I3av628q6dt6mq",
  "I7mf0sij342109",
  "Ibvirp862qkkup",
  "Ic5b47dj4coa3r",
  "Idehabrqi23sc0",
  "Ibs22tt76qp5bi",
  "I37r4bdai8o9mp",
  "Idn2ghub1o4i40",
  "I48u78djt89dod",
  "Ifmaahl40gom3g",
  "Idfpo6162k0hq",
  "Ib85m5kfbepu2t",
  "Idaml5bdhsfcsl",
  "I70iuri2ilha1f",
  "I19hvnphoaj44l",
  "I1ng31ej27mh4k",
  "Ieec0cu336gteb",
  "I815d5k4ij85nv",
  "Ifi98fgi9o46v7",
  "Ic0oj9tok33uap",
  "I3tdutpfjuk32j",
  "I1esdujrkdacpb",
  "I9g1d820jf9m2s",
  "I6gun5k9fbb4s0",
  "I85qkvekflgteq",
  "I1qt5nua7ua655",
  "I7cl9esn1l72m7",
  "Id68sq6o2gm8qi",
  "I8ve4g3egaln6a",
  "I6hnblkt2b5hd4",
  "If4hvqaeoqq5us",
  "Iafscmv8tjf0ou",
  "I2gpmmfdqv3cdc",
  "Ibtsa3docbr9el",
  "Iaiqv5prlisjkg",
  "XcmPalletQueryStatus",
  "Ic4qvh5df9s5gp",
  "I7vlvrrl2pnbgk",
  "I50sjs3s5lud21",
  "XcmPalletVersionMigrationStage",
  "I50qp0ij7h62g2",
  "Iteuj23is2ed5",
  "I3rp19gb4dadaa",
  "I9paqujeb1fpv6",
  "Iakevv83i18n4r",
  "If2ssl12kcglhg",
  "Iabk8ljl5g8c86",
  "Ic76kfh5ebqkpl",
  "Icrujen33bbibf",
  "I5gi8h3e5lkbeq",
  "Ibgm4rnf22lal1",
  "I8mmaab8je28oo",
  "I6r0pr82pbiftt",
  "I2aatv5i0cb96a",
  "Ib9msr5sr8t3dn",
  "I3le5tr7ugg6l2",
  "I3iun9sig164po",
  "I2uqmls7kcdnii",
  "Idg69klialbkb8",
  "I13jboebjcbglr",
  "I30pg328m00nr3",
  "I381dkhrurdhrs",
  "Ic8hi3qr11vngc",
  "Ibjdlecumfu7q7",
  "Ia9ems1kg7laoc",
  "I92fq0fa45vi3",
  "Id01dpp0dn2cj0",
  "I6nu8k62ck9o8o",
  "I6s1nbislhk619",
  "I260m120dp9sbk",
  "ParachainsInclusionAggregateMessageOrigin",
  "I53esa2ms463bk",
  "I1lfimt2mpej64",
  "Iasb8k6ash5mjn",
  "Ibv4ep0hngvn9e",
  "Ieoqregtp7b00",
  "I8uui7ldro956f",
  "I1tf93k54ltg1v",
  "I6ove5at7hfiur",
  "I9c0urppp07b8b",
  "VersionedLocatableAsset",
  "I6nmp4rhqla35",
  "I8ndstn7sf4kek",
  "I1a3ecmnlnvr59",
  "I2fb54desdqd9n",
  "I3pirohb0sp3ic",
  "Iemqna2uucuei9",
  "Idjett00s2gd",
]
`;

exports[`types-builder > imports > collects imports from both common types and papi itself 2`] = `
[
  "SS58String",
]
`;
